File: api/concept/type/AttributeType.ts
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import {Stream} from "../../../common/util/Stream";
import {TypeDBTransaction} from "../../connection/TypeDBTransaction";
import {Attribute} from "../thing/Attribute";
import {ThingType} from "./ThingType";
import {Concept} from "../Concept";
import {Value} from "../value/Value";
import {RequestBuilder} from "../../../common/rpc/RequestBuilder";
import Annotation = ThingType.Annotation;
import Transitivity = Concept.Transitivity;
import ValueType = Concept.ValueType;

export interface AttributeType extends ThingType {
    readonly valueType: ValueType;

    put(transaction: TypeDBTransaction, value: Value): Promise<Attribute>;
    putBoolean(transaction: TypeDBTransaction, value: boolean): Promise<Attribute>;
    putLong(transaction: TypeDBTransaction, value: number): Promise<Attribute>;
    putDouble(transaction: TypeDBTransaction, value: number): Promise<Attribute>;
    putString(transaction: TypeDBTransaction, value: string): Promise<Attribute>;
    putDateTime(transaction: TypeDBTransaction, value: Date): Promise<Attribute>;

    get(transaction: TypeDBTransaction, value: Value): Promise<Attribute>;
    getBoolean(transaction: TypeDBTransaction, value: boolean): Promise<Attribute>;
    getLong(transaction: TypeDBTransaction, value: number): Promise<Attribute>;
    getDouble(transaction: TypeDBTransaction, value: number): Promise<Attribute>;
    getString(transaction: TypeDBTransaction, value: string): Promise<Attribute>;
    getDateTime(transaction: TypeDBTransaction, value: Date): Promise<Attribute>;

    getSupertype(transaction: TypeDBTransaction): Promise<AttributeType>;
    setSupertype(transaction: TypeDBTransaction, type: AttributeType): Promise<void>;

    getSupertypes(transaction: TypeDBTransaction): Stream<AttributeType>;

    getSubtypes(transaction: TypeDBTransaction): Stream<AttributeType>;
    getSubtypes(transaction: TypeDBTransaction, valueType: ValueType): Stream<AttributeType>;
    getSubtypes(transaction: TypeDBTransaction, transitivity: Transitivity): Stream<AttributeType>;
    getSubtypes(transaction: TypeDBTransaction, valueType: ValueType, transitivity: Transitivity): Stream<AttributeType>;

    getInstances(transaction: TypeDBTransaction): Stream<Attribute>;
    getInstances(transaction: TypeDBTransaction, transitivity: Transitivity): Stream<Attribute>;

    getOwners(transaction: TypeDBTransaction): Stream<ThingType>;
    getOwners(transaction: TypeDBTransaction, annotations: Annotation[]): Stream<ThingType>;
    getOwners(transaction: TypeDBTransaction, transitivity: Transitivity): Stream<ThingType>;
    getOwners(transaction: TypeDBTransaction, annotations: Annotation[], transitivity: Transitivity): Stream<ThingType>;

    getRegex(transaction: TypeDBTransaction): Promise<string>;
    setRegex(transaction: TypeDBTransaction, regex: string): Promise<void>;
    unsetRegex(transaction: TypeDBTransaction): Promise<void>;
}

export namespace AttributeType {
    export function proto(attributeType: AttributeType) {
        return RequestBuilder.Type.AttributeType.protoAttributeType(attributeType.label);
    }
}

File: api/concept/type/AttributeType.java
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.vaticle.typedb.client.api.concept.type;

import com.vaticle.typedb.client.api.TypeDBTransaction;
import com.vaticle.typedb.client.api.concept.Value;
import com.vaticle.typedb.client.api.concept.thing.Attribute;
import com.vaticle.typedb.client.common.exception.TypeDBClientException;

import javax.annotation.CheckReturnValue;
import javax.annotation.Nullable;
import java.time.LocalDateTime;
import java.util.Set;
import java.util.stream.Stream;

import static com.vaticle.typedb.client.common.exception.ErrorMessage.Concept.BAD_ENUM_VALUE;

public interface AttributeType extends ThingType {
    @CheckReturnValue
    ValueType getValueType();

    @Override
    @CheckReturnValue
    default boolean isAttributeType() {
        return true;
    }

    @Override
    @CheckReturnValue
    default AttributeType asAttributeType() {
        return this;
    }

    Attribute put(TypeDBTransaction transaction, Value value);

    default Attribute put(TypeDBTransaction transaction, String value) {
        return put(transaction, Value.of(value));
    }

    default Attribute put(TypeDBTransaction transaction, long value) {
        return put(transaction, Value.of(value));
    }

    default Attribute put(TypeDBTransaction transaction, double value) {
        return put(transaction, Value.of(value));
    }

    default Attribute put(TypeDBTransaction transaction, boolean value) {
        return put(transaction, Value.of(value));
    }

    default Attribute put(TypeDBTransaction transaction, LocalDateTime value) {
        return put(transaction, Value.of(value));
    }

    @Nullable
    Attribute get(TypeDBTransaction transaction, Value value);

    @Nullable
    default Attribute get(TypeDBTransaction transaction, String value) {
        return get(transaction, Value.of(value));
    }

    @Nullable
    default Attribute get(TypeDBTransaction transaction, long value) {
        return get(transaction, Value.of(value));
    }

    @Nullable
    default Attribute get(TypeDBTransaction transaction, double value) {
        return get(transaction, Value.of(value));
    }

    @Nullable
    default Attribute get(TypeDBTransaction transaction, boolean value) {
        return get(transaction, Value.of(value));
    }

    @Nullable
    default Attribute get(TypeDBTransaction transaction, LocalDateTime value) {
        return get(transaction, Value.of(value));
    }

    String getRegex(TypeDBTransaction transaction);

    void setRegex(TypeDBTransaction transaction, String regex);

    void unsetRegex(TypeDBTransaction transaction);

    @CheckReturnValue
    default boolean isBoolean() {
        return getValueType() == ValueType.BOOLEAN;
    }

    @CheckReturnValue
    default boolean isLong() {
        return getValueType() == ValueType.LONG;
    }

    @CheckReturnValue
    default boolean isDouble() {
        return getValueType() == ValueType.DOUBLE;
    }

    @CheckReturnValue
    default boolean isString() {
        return getValueType() == ValueType.STRING;
    }

    @CheckReturnValue
    default boolean isDateTime() {
        return getValueType() == ValueType.DATETIME;
    }

    enum ValueType {
        OBJECT(Object.class, false, false),
        BOOLEAN(Boolean.class, true, false),
        LONG(Long.class, true, true),
        DOUBLE(Double.class, true, false),
        STRING(String.class, true, true),
        DATETIME(LocalDateTime.class, true, true);

        private final Class<?> valueClass;
        private final boolean isWritable;
        private final boolean isKeyable;

        ValueType(Class<?> valueClass, boolean isWritable, boolean isKeyable) {
            this.valueClass = valueClass;
            this.isWritable = isWritable;
            this.isKeyable = isKeyable;
        }

        @CheckReturnValue
        public static ValueType of(com.vaticle.typedb.client.jni.ValueType valueType) {
            switch (valueType) {
                case Boolean: return BOOLEAN;
                case Long: return LONG;
                case Double: return DOUBLE;
                case String: return STRING;
                case DateTime: return DATETIME;
                case Object: return OBJECT;
            }
            throw new TypeDBClientException(BAD_ENUM_VALUE);
        }

        @CheckReturnValue
        public com.vaticle.typedb.client.jni.ValueType asJNI(){
            switch (this) {
                case BOOLEAN: return com.vaticle.typedb.client.jni.ValueType.Boolean;
                case LONG: return com.vaticle.typedb.client.jni.ValueType.Long;
                case DOUBLE: return com.vaticle.typedb.client.jni.ValueType.Double;
                case STRING: return com.vaticle.typedb.client.jni.ValueType.String;
                case DATETIME: return com.vaticle.typedb.client.jni.ValueType.DateTime;
                case OBJECT: return com.vaticle.typedb.client.jni.ValueType.Object;
            }
            throw new TypeDBClientException(BAD_ENUM_VALUE);
        }

        @CheckReturnValue
        public Class<?> valueClass() {
            return valueClass;
        }

        @CheckReturnValue
        public boolean isWritable() {
            return isWritable;
        }

        @CheckReturnValue
        public boolean isKeyable() {
            return isKeyable;
        }
    }

    void setSupertype(TypeDBTransaction transaction, AttributeType attributeType);

    @Override
    @CheckReturnValue
    Stream<? extends AttributeType> getSubtypes(TypeDBTransaction transaction);

    @CheckReturnValue
    Stream<? extends AttributeType> getSubtypes(TypeDBTransaction transaction, ValueType valueType);

    @Override
    @CheckReturnValue
    Stream<? extends AttributeType> getSubtypesExplicit(TypeDBTransaction transaction);

    @Override
    @CheckReturnValue
    Stream<? extends Attribute> getInstances(TypeDBTransaction transaction);

    @Override
    @CheckReturnValue
    Stream<? extends Attribute> getInstancesExplicit(TypeDBTransaction transaction);

    @CheckReturnValue
    Stream<? extends ThingType> getOwners(TypeDBTransaction transaction);

    @CheckReturnValue
    Stream<? extends ThingType> getOwners(TypeDBTransaction transaction, Set<Annotation> annotations);

    @CheckReturnValue
    Stream<? extends ThingType> getOwnersExplicit(TypeDBTransaction transaction);

    @CheckReturnValue
    Stream<? extends ThingType> getOwnersExplicit(TypeDBTransaction transaction, Set<Annotation> annotations);
}
